md-proto/WebAssembly.md

```diff
@@ -0,0 +1,1693 @@
+# WebAssembly Specification
+
+0. [Module](#module)
+0. [Validation](#validation)
+0. [Execution](#execution)
+0. [Types](#types)
+0. [Signatures](#signatures)
+0. [Instruction Families](#instruction-families)
+0. [Instructions](#instructions)
+
+
+## Module
+
+TODO: describe the contents of a module
+
+### Positions within a function body
+
+Function bodies contain a sequence of instructions. A *position* within a
+function refers to an element of the sequence, or to an additional
+*return position* which is sequenced after the end of the function body.
+
+
+## Validation
+
+### Module Validation
+
+TODO: describe module validation
+
+### Function Validation
+
+For the duration of the validation of a function body, several data structures
+are created:
+ - A *control flow stack*, with entries containing:
+      - Either a [type](#types), `any` which satisfies any type check, or `void`
+        which satisfies no type check.
+      - A *depth* value.
+ - A *type stack*.
+ - A *current position*.
+
+During validation, if either stack is empty when an element is to be popped
+from it, validation fails.
+
+> `any` is effectively serving as a kind of bottom type, and `void` is
+effectively serving as a kind of top type. Note that in some type theory
+contexts, it's common to use these names for exactly opposite purposes.
+
+#### Type Stack Pop
+
+To perform a pop from the type stack:
+ - If the type stack's size is less than the control flow stack's top's depth
+   value, return `void`.
```

Comment:
It doesn't necessarily fail validation though. Also, in general, the validation code does need further work.